### ISSUE TITLE
Incorrect candidate list from candidates endpoint

### ISSUE THREAD

**Issue by jwchumley** [fecgov/openFEC#3673]
From Public Records - The elections endpoint for MI House District 03 is correct. The candidate/totals endpoint for the same criteria is not correct. It looks like several issues:

1. When you click on this (not URL has cycle=2020)
Page link: https://www.fec.gov/data/candidates/house/?cycle=2020&state=MI&district=03

The selected 2 year period is 2018 and doesn't match the URL. The list isn't quite right for 2018 but it is definitely wrong for 2020.

2. If you grab the API call and change it to 2020 like this:

API call: https://api.open.fec.gov/v1/candidates/totals/?api_key=DEMO_KEY&sort_hide_null=false&sort_nulls_last=true&cycle=2020&state=MI&district=03&sort=-receipts&per_page=30&page=1&office=H

It's still wrong:
H0MI03126 (good)
H6MI03172 (bad)

 but if you add &election_full=true then the list is correct.

H0MI03233
H0MI03258
H0MI03241
H8MI03079
H0MI03126

**Comment by jwchumley** [comment]
Here are more examples:

https://www.fec.gov/data/elections/house/NC/03/2020/

('H2NC13078','H0NC03149','H0NC03156')

**Comment by lbeaufort** [comment]
This is a sqlalchemy-heavy endpoint. 

```
SELECT DISTINCT anon_1.candidate_id, anon_1.candidate_election_year, anon_1.candidate_name, anon_1.party_full, anon_1.incumbent_challenge_full, anon_1.office, anon_1.total_receipts, anon_1.total_disbursements, anon_1.cash_on_hand_end_period, anon_1.committee_ids, anon_1.coverage_end_date, anon_1.candidate_pcc_id, ofec_committee_detail_mv.name AS candidate_pcc_name 
FROM (SELECT anon_2.candidate_id AS candidate_id, anon_2.candidate_election_year AS candidate_election_year, max(anon_2.candidate_name) AS candidate_name, max(anon_2.party_full) AS party_full, max(anon_2.incumbent_challenge_full) AS incumbent_challenge_full, max(anon_2.office) AS office, sum(coalesce(anon_2.receipts, 0.0)) AS total_receipts, sum(coalesce(anon_2.disbursements, 0.0)) AS total_disbursements, sum(coalesce(anon_2.last_cash_on_hand_end_period, 0.0)) AS cash_on_hand_end_period, array_agg(DISTINCT anon_2.committee_id) AS committee_ids, max(anon_2.coverage_end_date) AS coverage_end_date, max(anon_2.candidate_pcc_id) AS candidate_pcc_id 
FROM (SELECT anon_3.candidate_id AS candidate_id, anon_3.candidate_election_year AS candidate_election_year, anon_3.committee_id AS committee_id, max(anon_3.name) AS candidate_name, max(anon_3.party_full) AS party_full, max(anon_3.incumbent_challenge_full) AS incumbent_challenge_full, max(anon_3.office) AS office, sum(coalesce(anon_3.receipts, 0.0)) AS receipts, sum(coalesce(anon_3.disbursements, 0.0)) AS disbursements, max(coalesce(anon_3.last_cash_on_hand_end_period, 0.0)) AS last_cash_on_hand_end_period, max(anon_3.coverage_end_date) AS coverage_end_date, max(anon_3.candidate_pcc_id) AS candidate_pcc_id 
FROM (SELECT anon_4.candidate_id AS candidate_id, anon_4.name AS name, anon_4.party_full AS party_full, anon_4.incumbent_challenge_full AS incumbent_challenge_full, anon_4.office AS office, anon_4.two_year_period AS two_year_period, anon_4.candidate_election_year AS candidate_election_year, anon_4.committee_id AS committee_id, anon_4.receipts AS receipts, anon_4.disbursements AS disbursements, first_value(anon_4.last_cash_on_hand_end_period) OVER (PARTITION BY anon_4.candidate_id, anon_4.candidate_election_year, anon_4.committee_id ORDER BY CASE WHEN (anon_4.coverage_end_date IS NOT NULL) THEN anon_4.two_year_period END DESC NULLS LAST) AS last_cash_on_hand_end_period, first_value(anon_4.coverage_end_date) OVER (PARTITION BY anon_4.candidate_id, anon_4.candidate_election_year, anon_4.committee_id ORDER BY CASE WHEN (anon_4.coverage_end_date IS NOT NULL) THEN anon_4.two_year_period END DESC NULLS LAST) AS coverage_end_date, anon_4.candidate_pcc_id AS candidate_pcc_id 
FROM (SELECT DISTINCT ofec_candidate_history_mv.candidate_id AS candidate_id, ofec_candidate_history_mv.name AS name, ofec_candidate_history_mv.party_full AS party_full, ofec_candidate_history_mv.incumbent_challenge_full AS incumbent_challenge_full, ofec_candidate_history_mv.office AS office, ofec_candidate_history_mv.two_year_period AS two_year_period, ofec_candidate_history_mv.candidate_election_year AS candidate_election_year, ofec_cand_cmte_linkage_mv.cmte_id AS committee_id, ofec_totals_house_senate_mv.receipts AS receipts, ofec_totals_house_senate_mv.disbursements AS disbursements, ofec_totals_house_senate_mv.last_cash_on_hand_end_period AS last_cash_on_hand_end_period, ofec_totals_house_senate_mv.coverage_end_date AS coverage_end_date, CASE WHEN (ofec_cand_cmte_linkage_mv.cmte_dsgn = 'P') THEN ofec_cand_cmte_linkage_mv.cmte_id END AS candidate_pcc_id 
FROM ofec_candidate_history_mv LEFT OUTER JOIN ofec_cand_cmte_linkage_mv ON ofec_candidate_history_mv.candidate_id = ofec_cand_cmte_linkage_mv.cand_id AND ofec_candidate_history_mv.two_year_period = ofec_cand_cmte_linkage_mv.fec_election_yr LEFT OUTER JOIN ofec_totals_house_senate_mv ON ofec_cand_cmte_linkage_mv.cmte_id = ofec_totals_house_senate_mv.committee_id AND ofec_cand_cmte_linkage_mv.fec_election_yr = ofec_totals_house_senate_mv.cycle AND ofec_cand_cmte_linkage_mv.cmte_dsgn IN ('P', 'A') 
WHERE ofec_cand_cmte_linkage_mv.cmte_dsgn IN ('P', 'A') AND ofec_candidate_history_mv.two_year_period <= 2020 AND ofec_candidate_history_mv.two_year_period > 2018 AND ofec_candidate_history_mv.candidate_election_year + ofec_candidate_history_mv.candidate_election_year % 2 = 2020 AND ofec_candidate_history_mv.office = 'H' AND ofec_candidate_history_mv.state = 'NC' AND ofec_candidate_history_mv.district = '03' AND ofec_candidate_history_mv.candidate_inactive = false) AS anon_4) AS anon_3 GROUP BY anon_3.candidate_id, anon_3.candidate_election_year, anon_3.committee_id) AS anon_2 GROUP BY anon_2.candidate_id, anon_2.candidate_election_year) AS anon_1 LEFT OUTER JOIN ofec_committee_detail_mv ON anon_1.candidate_pcc_id = ofec_committee_detail_mv.committee_id
```

**Comment by lbeaufort** [comment]
@jwchumley the election profile page is fed by a different endpoint (/elections/) from the candidate data table (candidates/totals/)- I think that should be a different bug. cc: @PaulClark2

**Comment by PaulClark2** [comment]
@lbeaufort @jwchumley  if it's from a different endpoint, we should log it as a separate bug.

**Comment by jwchumley** [comment]
Agreed. I thought it was the same but it isn't.

**Comment by jwchumley** [comment]
Ignore my second list of three candidates. They are fine. See edit above.

**Comment by lbeaufort** [comment]
After speaking with @jwchumley, the candidate lists for the candidate data tables, candidate totals data tables, and election profile pages should all match.

**Comment by lbeaufort** [comment]
@fecjjeng is doing a major overhaul of /candidates/totals/ that should be done before taking a closer look.

**Comment by lbeaufort** [comment]
The overhaul of candidates/totals has been completed with https://github.com/fecgov/openFEC/pull/3726 so we are good to work on this issue

**Comment by lbeaufort** [comment]
@fecjjeng has a PR in to make more changes to candidate totals, so test with those changes first.

**Comment by lbeaufort** [comment]
From @jwchumley 

Truth:   
This also picks up off year election candidates: 
```
select * from select * from DISCLOSURE.CAND_VALID_FEC_YR
where fec_election_yr=2020
and cand_election_yr in (fec_election_yr-1,fec_election_yr)
and cand_state='MI'
and cand_office_district='03'
```

**Comment by lbeaufort** [comment]
Update - it looks like `election_full=True` has the _more_ correct results, but both /elections/ and /candidates/totals/ are missing one candidate, even with `election_full` set to `True`. I think the best path forward is: 
1) update the front end house candidates total to pass `election_full=True`, relabel time frame from two-year-period to election year to match other data tables 
2) investigate the missing candidate from both /elections/ and /candidates/totals/ 
3) look to make `election_full=True` match `election_full=False` for house candidates

**Comment by lbeaufort** [comment]
LOWER, JAMES (JIM) AARON   | H0MI03274

**Comment by jason-upchurch** [comment]
@lbeaufort as part of the related test case, I saw that `SMITH, DOUGLAS MARK MR. JR.` appears in the 2020 version of the very first link in this issue (3 results), but not in the truth query for 2020 (6 results). The reason for this is that for this candidate, `cand_election_yr`=2018, but to be included in the result, `cand_election_yr` must be in [2019, 2020].

I've updated #3791 with a failing test where `fec_election_yr`=2018, and I set `cand_election_yr`s at 2017, 2018, and 2020. As expected, the query returns 2 results, but the api call returns 3 results (assuming you wanted `fec_election_yr`=2020).

I've also noticed that there are 19,224 cases where `fec_election_yr` exceeds `cand_election_yr` by 3 or more, and 1,310 cases where `cand_election_yr` exceeds `fec_election_yr` by 3 or more. This may be relevant to queries requiring `cand_election_yr` in (`fec_election_yr`-1, `fec_election_yr`).

```sql
select * from disclosure.cand_valid_fec_yr where cand_id = 'H6MI03172' and fec_election_yr=2020;
```

**Comment by fecjjeng** [comment]
@jason-upchurch Your observation is correct.  And that is exactly the reason we have to do a heavy lifting in the logic of cand_election_yr in issue #3736 (related to problem discovered in Issue #3700) is that the cand_election_yr, and the cand_election_yr/fec_electionn_yr relation  in disclosure.cand_valid_fec_yr and disclosure.cand_cmte_linkage tables does not fit the way we use use these information.  My understanding from Paul/Jeff is that, as long as the the candidate committee(s) still existing and filing financial data, a new row with the previous cand_election_yr will be created every fec_election_yr.  That makes the cand_election_yr data not the "real" cand_election_yr.   The problem started getting hairy when candidate has gaps between his/her elections.  Hillary Clinton is a good example.  That is, a candidate list is not really meaningful based on fec_election_yr, a candidate based on cand_election_yr is more correct.

**Comment by lbeaufort** [comment]
@jason-upchurch re: SMITH, DOUGLAS MARK MR. JR. if the `cand_election_yr` is 2018 then I wouldn't expect them to show up in the 2020 list if the 2020 list is supposed to be a list of who is running in 2020.

Here's the new query I'm running on the front end, which uses election year instead of two-year period: 
https://fec-dev-api.app.cloud.gov/v1/candidates/totals/?api_key=DEMO_KEY&sort_hide_null=false&sort_nulls_last=true&election_year=2020&election_full=True&state=MI&district=03&sort=-receipts&per_page=30&page=1&office=H

**Comment by jason-upchurch** [comment]
@lbeaufort agreed! I just wanted to note that this candidate appears here for 2020 but this may be a separate issue: https://www.fec.gov/data/candidates/house/?cycle=2020&state=MI&district=03.

**Comment by lbeaufort** [comment]
@jason-upchurch thank you - I think with the proposed change to the front-end (https://github.com/fecgov/fec-cms/pull/2917) this shouldn't be an issue anymore.

**Comment by lbeaufort** [comment]
Update: 

✅ 1) update the front end house candidates total to pass `election_full=True`, relabel time frame from two-year-period to election year to match other data tables >> PR ready for review fecgov/fec-cms#2917) 
✅ 2) investigate the missing candidate from both /elections/ and /candidates/totals/  >> no longer missing from /elections/ - we will find more examples when we cross-compare the endpoints (#3719)
3) look to make `election_full=True` match `election_full=False` for house candidates >> I think this is lower-priority than #3796 so I will split this into another issue

**Comment by lbeaufort** [comment]
PR has been merged
